File: scripts/chart-usage.py
#!/usr/bin/env python3
"""
Render a stacked-bar SVG of daily output tokens by model family from
exports/daily_by_model.csv (produced by export-usage-csv.sh).

No third-party dependencies — emits plain SVG you can open in any browser.

Usage: python3 scripts/chart-usage.py [exports/daily_by_model.csv] [exports/usage-trend.svg]
"""
import csv
import sys
from collections import defaultdict

IN = sys.argv[1] if len(sys.argv) > 1 else "exports/daily_by_model.csv"
OUT = sys.argv[2] if len(sys.argv) > 2 else "exports/usage-trend.svg"

# model family -> color
COLORS = {
    "opus": "#b45ad6",
    "sonnet": "#0a84ff",
    "haiku": "#28a745",
    "other": "#888888",
}


def family(model: str) -> str:
    m = model.lower()
    for fam in ("opus", "sonnet", "haiku"):
        if fam in m:
            return fam
    return "other"


# date -> family -> output tokens
data: dict[str, dict[str, float]] = defaultdict(lambda: defaultdict(float))
with open(IN, newline="") as f:
    for row in csv.DictReader(f):
        data[row["date"]][family(row["model"])] += float(row["output_tokens"])

dates = sorted(data)
if not dates:
    sys.exit("no data in " + IN)
families = [fam for fam in ("opus", "sonnet", "haiku", "other")
           if any(data[d].get(fam) for d in dates)]

# ── layout ────────────────────────────────────────────────────────────────────
W, H = 1100, 480
ML, MR, MT, MB = 70, 20, 40, 90
plot_w, plot_h = W - ML - MR, H - MT - MB
bar_w = plot_w / len(dates) * 0.8
gap = plot_w / len(dates)

max_total = max(sum(data[d].values()) for d in dates) or 1
# round axis up to a nice number
import math
step = 10 ** math.floor(math.log10(max_total))
axis_max = math.ceil(max_total / step) * step


def y(v):  # value -> svg y
    return MT + plot_h - (v / axis_max) * plot_h


def human(v):
    for unit, div in (("B", 1e9), ("M", 1e6), ("K", 1e3)):
        if v >= div:
            return f"{v/div:.1f}{unit}"
    return str(int(v))


svg = [f'<svg xmlns="http://www.w3.org/2000/svg" width="{W}" height="{H}" font-family="-apple-system,Helvetica,Arial,sans-serif">']
svg.append(f'<rect width="{W}" height="{H}" fill="white"/>')
svg.append(f'<text x="{ML}" y="24" font-size="16" font-weight="bold">Claude Code — daily output tokens by model</text>')

# y gridlines + labels
ticks = 5
for i in range(ticks + 1):
    val = axis_max * i / ticks
    yy = y(val)
    svg.append(f'<line x1="{ML}" y1="{yy:.1f}" x2="{W-MR}" y2="{yy:.1f}" stroke="#eee"/>')
    svg.append(f'<text x="{ML-8}" y="{yy+4:.1f}" font-size="11" fill="#666" text-anchor="end">{human(val)}</text>')

# bars
for i, d in enumerate(dates):
    x = ML + i * gap + (gap - bar_w) / 2
    y_cursor = y(0)
    for fam in families:
        v = data[d].get(fam, 0)
        if v <= 0:
            continue
        h = (v / axis_max) * plot_h
        y_cursor -= h
        svg.append(f'<rect x="{x:.1f}" y="{y_cursor:.1f}" width="{bar_w:.1f}" height="{h:.1f}" fill="{COLORS[fam]}"/>')
    # x label (rotated) every Nth to avoid crowding
    if len(dates) <= 40 or i % 2 == 0:
        lx = x + bar_w / 2
        svg.append(f'<text x="{lx:.1f}" y="{MT+plot_h+14:.1f}" font-size="9" fill="#666" text-anchor="end" transform="rotate(-60 {lx:.1f} {MT+plot_h+14:.1f})">{d[5:]}</text>')

# legend
lx = ML
ly = H - 24
for fam in families:
    svg.append(f'<rect x="{lx}" y="{ly-10}" width="12" height="12" fill="{COLORS[fam]}"/>')
    svg.append(f'<text x="{lx+18}" y="{ly}" font-size="12" fill="#333">{fam}</text>')
    lx += 90

svg.append("</svg>")

with open(OUT, "w") as f:
    f.write("\n".join(svg))
print(f"wrote {OUT} ({len(dates)} days, families: {', '.join(families)})")

File: scripts/chart-usage.cs
#!/usr/bin/env dotnet
// Render a stacked-bar SVG of daily output tokens by model family from
// exports/daily_by_model.csv (produced by export-usage-csv.sh).
//
// Usage: dotnet run scripts/chart-usage.cs [in.csv] [out.svg]

using System.Globalization;
using System.Text;

var input = args.Length > 0 ? args[0] : "exports/daily_by_model.csv";
var output = args.Length > 1 ? args[1] : "exports/usage-trend.svg";

// Model family -> colour. Iteration order fixes the stacking order, and the
// legend is derived from this too, so a family added here can't be counted
// into the data while silently missing from the chart.
var colors = new (string Family, string Hex)[]
{
    ("fable", "#e8833a"),
    ("opus", "#b45ad6"),
    ("sonnet", "#0a84ff"),
    ("haiku", "#28a745"),
    ("other", "#888888"),
};

// Checked in order, so a model matching two names lands in the first.
string[] familyNames = ["fable", "mythos", "opus", "sonnet", "haiku"];

// Not static, so it reads the one list above rather than repeating it.
string FamilyOf(string model)
{
    var m = model.ToLowerInvariant();
    foreach (var fam in familyNames)
    {
        if (m.Contains(fam, StringComparison.Ordinal))
        {
            // Mythos shares Fable's rate and colour.
            return fam == "mythos" ? "fable" : fam;
        }
    }
    return "other";
}

// date -> family -> output tokens
var data = new SortedDictionary<string, Dictionary<string, double>>(StringComparer.Ordinal);

foreach (var row in ReadCsv(input))
{
    if (!row.TryGetValue("date", out var date) ||
        !row.TryGetValue("model", out var model) ||
        !row.TryGetValue("output_tokens", out var tokensText))
    {
        continue;
    }
    if (!double.TryParse(tokensText, NumberStyles.Float, CultureInfo.InvariantCulture, out var tokens))
    {
        continue;
    }

    if (!data.TryGetValue(date, out var byFamily))
    {
        byFamily = [];
        data[date] = byFamily;
    }
    var family = FamilyOf(model);
    byFamily[family] = byFamily.GetValueOrDefault(family) + tokens;
}

var dates = data.Keys.ToList();
if (dates.Count == 0)
{
    Console.Error.WriteLine($"no data in {input}");
    return 1;
}

var families = colors
    .Where(c => dates.Any(d => data[d].GetValueOrDefault(c.Family) > 0))
    .ToList();

// ── layout ───────────────────────────────────────────────────────────────────
const int W = 1100, H = 480;
const int ML = 70, MR = 20, MT = 40, MB = 90;
const double PlotW = W - ML - MR, PlotH = H - MT - MB;

var gap = PlotW / dates.Count;
var barW = gap * 0.8;

var maxTotal = dates.Max(d => data[d].Values.Sum());
if (maxTotal <= 0) maxTotal = 1;

// Round the axis up to a nice number.
var step = Math.Pow(10, Math.Floor(Math.Log10(maxTotal)));
var axisMax = Math.Ceiling(maxTotal / step) * step;

double Y(double v) => MT + PlotH - (v / axisMax) * PlotH;

static string Human(double v)
{
    foreach (var (unit, div) in ((string, double)[])[("B", 1e9), ("M", 1e6), ("K", 1e3)])
    {
        if (v >= div) return F(v / div, "0.0") + unit;
    }
    return ((long)v).ToString(CultureInfo.InvariantCulture);
}

// Every number written into the SVG goes through here. A locale that uses a
// comma as the decimal separator would otherwise emit coordinates the renderer
// silently rejects.
static string F(double v, string format = "0.0") =>
    v.ToString(format, CultureInfo.InvariantCulture);

var svg = new StringBuilder();
svg.AppendLine($"""<svg xmlns="http://www.w3.org/2000/svg" width="{W}" height="{H}" font-family="-apple-system,Helvetica,Arial,sans-serif">""");
svg.AppendLine($"""<rect width="{W}" height="{H}" fill="white"/>""");
svg.AppendLine($"""<text x="{ML}" y="24" font-size="16" font-weight="bold">Claude Code — daily output tokens by model</text>""");

const int Ticks = 5;
for (var i = 0; i <= Ticks; i++)
{
    var val = axisMax * i / Ticks;
    var yy = Y(val);
    svg.AppendLine($"""<line x1="{ML}" y1="{F(yy)}" x2="{W - MR}" y2="{F(yy)}" stroke="#eee"/>""");
    svg.AppendLine($"""<text x="{ML - 8}" y="{F(yy + 4)}" font-size="11" fill="#666" text-anchor="end">{Human(val)}</text>""");
}

for (var i = 0; i < dates.Count; i++)
{
    var date = dates[i];
    var x = ML + i * gap + (gap - barW) / 2;
    var cursor = Y(0);

    foreach (var (family, hex) in families)
    {
        var v = data[date].GetValueOrDefault(family);
        if (v <= 0) continue;
        var h = (v / axisMax) * PlotH;
        cursor -= h;
        svg.AppendLine($"""<rect x="{F(x)}" y="{F(cursor)}" width="{F(barW)}" height="{F(h)}" fill="{hex}"/>""");
    }

    // Rotated x label, thinned out once the bars get crowded.
    if (dates.Count <= 40 || i % 2 == 0)
    {
        var lx = x + barW / 2;
        var ly = MT + PlotH + 14;
        var label = date.Length > 5 ? date[5..] : date;
        svg.AppendLine($"""<text x="{F(lx)}" y="{F(ly)}" font-size="9" fill="#666" text-anchor="end" transform="rotate(-60 {F(lx)} {F(ly)})">{label}</text>""");
    }
}

var legendX = ML;
const int LegendY = H - 24;
foreach (var (family, hex) in families)
{
    svg.AppendLine($"""<rect x="{legendX}" y="{LegendY - 10}" width="12" height="12" fill="{hex}"/>""");
    svg.AppendLine($"""<text x="{legendX + 18}" y="{LegendY}" font-size="12" fill="#333">{family}</text>""");
    legendX += 90;
}

svg.Append("</svg>");

File.WriteAllText(output, svg.ToString());
Console.WriteLine($"wrote {output} ({dates.Count} days, families: {string.Join(", ", families.Select(f => f.Family))})");
return 0;

// Minimal RFC 4180 reader: the export quotes its string columns, so bare
// splitting on commas would break on any quoted field.
static IEnumerable<Dictionary<string, string>> ReadCsv(string path)
{
    using var reader = new StreamReader(path);
    var header = reader.ReadLine();
    if (header is null) yield break;

    var columns = SplitCsvLine(header);
    string? line;
    while ((line = reader.ReadLine()) is not null)
    {
        if (line.Length == 0) continue;
        var fields = SplitCsvLine(line);
        var row = new Dictionary<string, string>(StringComparer.Ordinal);
        for (var i = 0; i < columns.Count && i < fields.Count; i++)
        {
            row[columns[i]] = fields[i];
        }
        yield return row;
    }
}

static List<string> SplitCsvLine(string line)
{
    var fields = new List<string>();
    var field = new StringBuilder();
    var quoted = false;

    for (var i = 0; i < line.Length; i++)
    {
        var c = line[i];
        if (quoted)
        {
            if (c == '"')
            {
                // A doubled quote inside a quoted field is a literal quote.
                if (i + 1 < line.Length && line[i + 1] == '"') { field.Append('"'); i++; }
                else quoted = false;
            }
            else field.Append(c);
        }
        else if (c == '"') quoted = true;
        else if (c == ',') { fields.Add(field.ToString()); field.Clear(); }
        else field.Append(c);
    }

    fields.Add(field.ToString());
    return fields;
}
